doc/examples/timeouts.rst

```diff
@@ -0,0 +1,161 @@
+          # a timeout exception.
+          coll.find_one({"name": "Nunu"})
+  except PyMongoError as exc:
+      if exc.timeout:
+          print(f"block timed out: {exc!r}")
+      else:
+          print(f"failed with non-timeout error: {exc!r}")
+
+The :attr:`pymongo.errors.PyMongoError.timeout` property (added in PyMongo 4.2)
+will be ``True`` when the error was caused by a timeout and ``False`` otherwise.
+
+The timeoutMS URI option
+------------------------
+
+PyMongo 4.2 also added support for the ``timeoutMS`` URI and keyword argument to
+:class:`~pymongo.mongo_client.MongoClient`. When this option is configured, the
+client will automatically apply the timeout to each API call. For example::
+
+  client = MongoClient("mongodb://localhost/?timeoutMS=10000")
+  coll = client.test.test
+  coll.insert_one({"name": "Nunu"})  # Uses a 10-second timeout.
+  coll.find_one({"name": "Nunu"})  # Also uses a 10-second timeout.
+
+The above is roughly equivalent to::
+
+  client = MongoClient()
+  coll = client.test.test
+  with pymongo.timeout(10):
+      coll.insert_one({"name": "Nunu"})
+  with pymongo.timeout(10):
+      coll.find_one({"name": "Nunu"})
+
+pymongo.timeout overrides timeoutMS
+-----------------------------------
+
+:meth:`~pymongo.timeout` overrides ``timeoutMS``; within a
+:meth:`~pymongo.timeout` block a client's ``timeoutMS`` option is ignored::
+
+  client = MongoClient("mongodb://localhost/?timeoutMS=10000")
+  coll = client.test.test
+  coll.insert_one({"name": "Nunu"})  # Uses the client's 10-second timeout.
+  # pymongo.timeout overrides the client's timeoutMS.
+  with pymongo.timeout(20):
+      coll.insert_one({"name": "Nunu"})  # Uses the 20-second timeout.
+  with pymongo.timeout(5):
+      coll.find_one({"name": "Nunu"})  # Uses the 5-second timeout.
+
+pymongo.timeout is thread safe
+------------------------------
+
+:meth:`~pymongo.timeout` is thread safe; the timeout only applies to current
+thread and multiple threads can configure different timeouts in parallel.
+
+pymongo.timeout is asyncio safe
+-------------------------------
+
+:meth:`~pymongo.timeout` is asyncio safe; the timeout only applies to current
+Task and multiple Tasks can configure different timeouts in parallel.
+:meth:`~pymongo.timeout` can be used identically in
+`Motor <https://github.com/mongodb/motor>`_, for example::
+
+  import motor.motor_asyncio
+  client = motor.motor_asyncio.AsyncIOMotorClient()
+  coll = client.test.test
+  with pymongo.timeout(10):
+      await coll.insert_one({"name": "Nunu"})
+      await coll.find_one({"name": "Nunu"})
+
+Troubleshooting
+---------------
+
+There are many timeout errors that can be raised depending on when the timeout
+expires. In code, these can be identified with the :attr:`pymongo.errors.PyMongoError.timeout`
+property. Some specific timeout errors examples are described below.
+
+The following error means that the client was unable to find an available server
```

Comment:
I would replace "The following error means that" with "When" for better flow in all of these examples

Reply:
Fixed.